crates/rpc/src/handlers/validator.rs

```diff
@@ -103,3 +110,54 @@ pub async fn validator_status(validator: &Validator, db: &ReamDB) -> Result<Stri
         Ok("active_ongoing".to_string())
     }
 }
+
+pub async fn get_validator_balances_from_state(
+    state_id: ID,
+    query: ValidatorBalanceQuery, 
+    db: ReamDB,
+) -> Result<impl Reply, Rejection> {
+
+    let state = get_state_from_id(state_id, &db).await?;
+
+    let filter: Option<HashSet<String>> = match query.id {
+        Some(ref ids) if ids.is_empty() => None,
+        Some(ids) => Some(ids.into_iter().collect()),
+        None => None,
+    };
+
+    //need to decide on the limit
+    if let Some(ref ids) = filter {
+        if ids.len() > 1000 {
+            return Err(ApiError::TooManyValidatorIds("Too many validator IDs in request".to_string()))?;
+        }
+    }
+
+    let validator_balances: Vec<ValidatorBalance> = state
+        .validators
+        .iter()
+        .enumerate()
+        .filter_map(|(i, validator)| {
+            if let Some(ref ids) = filter {
+                if !ids.contains(&i.to_string()) && !ids.contains(&format!("{:?}", validator.pubkey)) {
+                    return None;
+                }
+            }
+            let balance = state.balances.get(i).unwrap_or(&0);
+            Some(ValidatorBalance {
+                index: i.to_string(),
+                balance: balance.to_string(),
+            })
+        })
+        .collect();
+
+    let response = BeaconResponse {
+        execution_optimistic: false,
+        finalized: false,
+        data: validator_balances,
+    };
+
+    Ok(warp::reply::with_status(
+        warp::reply::json(&response),
+        StatusCode::OK,
+    ))
+}
```

Comment:
![image](https://github.com/user-attachments/assets/35519cd7-2624-4327-8e6d-cb02907147ca)

you are missing a new line